### PR TITLE
Give docs a proper update

### PR DESCRIPTION
**Pull request by jonhoo**
I've made a pass on pretty much all the docs, as well as the names of a few types, to try to clean up the externally-visible API. It'd be awesome if someone could give it all a read through and see that a) the structure of the API makes sense, b) I'm not saying anything totally false, and c) there aren't any gaping holes that we should fill immediately.

I have included any example code somewhat intentionally since I'm guessing the external APIs are likely to change a decent amount with #30.

### CODE REVIEW

File: src/flamegraph/attrs.rs [hunk 1]
```diff
@@ -18,17 +18,19 @@ macro_rules! unwrap_or_continue {
 pub struct FuncFrameAttrsMap(HashMap<String, FrameAttrs>);
 
 impl FuncFrameAttrsMap {
-    /// Parse a FuncFrameAttrsMap from a file.
-    /// Each line should be a function name followed by a tab,
-    /// then a sequence of tab separated name=value pairs.
+    /// Parse frame attribtues from a file.
```
Comment (jasonrhansen) [inline src/flamegraph/attrs.rs]: Typo: attribtues

File: src/flamegraph/attrs.rs [hunk 2]
```diff
@@ -18,17 +18,19 @@ macro_rules! unwrap_or_continue {
...
+    /// tab-separated `name=value` pairs.
     pub fn from_file(path: &PathBuf) -> io::Result<FuncFrameAttrsMap> {
         let file = BufReader::new(File::open(path)?);
         FuncFrameAttrsMap::from_reader(file)
     }
 
-    /// Parse a FuncFrameAttrsMap from a reader.
-    /// Each line should be a function name followed by a tab,
-    /// then a sequence of tab separated name=value pairs.
+    /// Parse frame attribtues from a `BufRead`.
```
Comment (jasonrhansen) [inline src/flamegraph/attrs.rs]: Same typo

File: src/flamegraph/mod.rs [hunk 3]
```diff
@@ -32,23 +28,67 @@ const XPAD: usize = 10; // pad lefm and right
...
+    /// Choose names based on the hashes of function names.
+    ///
+    /// This will cause similar functions to be colored similarly.
     pub hash: bool,
+
+    /// Store the choice of color for each function so that later invocations use the same colors.
+    ///
+    /// With this option enabled, a file called `palette.map` will be created the first time a
+    /// flame graph is generated, and the (random) color chosen for each function will be written
+    /// into it. On subsequent invocations, functions that already have a colored registered in
```
Comment (jasonrhansen) [inline src/flamegraph/mod.rs]: Change "colored" to "color".

File: src/flamegraph/mod.rs [hunk 4]
```diff
@@ -410,6 +463,12 @@ where
     from_sorted_lines(opt, input.lines(), writer)
 }
 
+/// Produce a flame graph from a set of reader that contain folded stack lines.
```
Comment (jasonrhansen) [inline src/flamegraph/mod.rs]: Change to "set of readers"

File: src/lib.rs [hunk 5]
```diff
@@ -1,9 +1,109 @@
...
+//! spending its time by looking at which functions appear in many of the traces. To ease this
+//! analysis, we want to "collapse" the stack traces so if a particular trace occurs more than
+//! once, we instead just keep it _once_ along with a count of how many times we've seen it. This
+//! is what the various collapsing tools do! You'll sometimes see the resulting tuples of stack +
+//! count called a "folded stack trace".
+//!
+//! Since profiling tools produce stack traces in a myriad of different formats, and the flame
+//! graph plotter expects input in a particular folded stack trace format, each profiler needs a
+//! separate collapse implementation. While the original Perl implementation supports _lots_ of
+//! profilers, Inferno currenly only supports one: the widely used [`perf`] tool (specifically the
```
Comment (jasonrhansen) [inline src/lib.rs]: Change "currenly" to "currently".

**Comment by jonhoo** [comment]
@jasonrhansen great, thanks for the review! I'll leave this open for a bit longer in case @JordiChauzi also has a chance to take a look, otherwise I'll merge towards the end of today EST.

File: src/flamegraph/color/mod.rs [hunk 6]
```diff
@@ -16,49 +16,88 @@ const BLUE_GRADIENT: (&str, &str) = ("#eeeeee", "#e0e0ff");
...
+/// different function names (kernel functions, JIT functions, etc.).
 #[derive(Clone, Copy, Debug, PartialEq)]
 pub enum MultiPalette {
+    /// Use Java semantics to color frames.
     Java,
+    /// Use JavaScript semantics to color frames.
     Js,
+    /// Use Perl semantics to color frames.
     Perl,
+    /// TODO: this seems to just be [`BasicPalette::Aqua`]?
```
Comment (JordiChauzi) [inline src/flamegraph/color/mod.rs]: The difference between `BasicPalette::Aqua` and `MultiPalette::Wakeup` is the default background gradient (`Grey` for `Aqua`, `Blue` for `Wakeup`).

File: src/flamegraph/mod.rs [hunk 7]
```diff
@@ -32,23 +28,67 @@ const XPAD: usize = 10; // pad lefm and right
...
+
+    /// Choose names based on the hashes of function names.
+    ///
+    /// This will cause similar functions to be colored similarly.
     pub hash: bool,
+
+    /// Store the choice of color for each function so that later invocations use the same colors.
+    ///
+    /// With this option enabled, a file called `palette.map` will be created the first time a
+    /// flame graph is generated, and the (random) color chosen for each function will be written
```
Comment (JordiChauzi) [inline src/flamegraph/mod.rs]: Do we need to precise random ? With `--hash` the color is always deterministic.

File: src/flamegraph/mod.rs [hunk 8]
```diff
@@ -32,23 +28,67 @@ const XPAD: usize = 10; // pad lefm and right
...
+    /// The title for the flame chart.
     pub title: String,
 }
 
+/// The direction the plot should grow.
 #[derive(Debug, Clone, Copy, Eq, PartialEq)]
 pub enum Direction {
+    /// Stacks grow from the bottom to the top.
+    ///
+    /// `main` will be at the bottom.
```
Comment (JordiChauzi) [inline src/flamegraph/mod.rs]: `main` might be too language specific, and even for language with `main`, it is possible to have a stack with no `main` (launching a `Thread` in java for example). Maybe change to `entry point` ?

**Comment by JordiChauzi** [comment]
For different structures implementing `Default`, you precised the default value in the doc of the structure. Hopefully the value will not change overtime, but is it not redundant ?

**Comment by jonhoo** [comment]
I think it's useful to spell out the defaults in the docs since they're not otherwise visible anywhere beyond digging into the code. It's true that we'd have to make sure update the docs if the defaults change, but at the same time I think these defaults should also be relatively unlikely to change :)

**Comment by jonhoo** [comment]
Thanks for the review from both of you! I'll merge once CI passes!

**Comment by jonhoo** [comment]
@jasonrhansen I added some more docs following #60 in 3e706ea6ee8ceaca22f716e0bec311020d1fb0e6. Can you take a quick look?

File: src/flamegraph/color/mod.rs [hunk 9]
```diff
@@ -16,49 +16,98 @@ const BLUE_GRADIENT: (&str, &str) = ("#eeeeee", "#e0e0ff");
...
+}
+
+/// A plain color palette in which the color is independent of the function name.
 #[derive(Clone, Copy, Debug, PartialEq)]
 pub enum BasicPalette {
+    /// A palette in which colors are chosen randomly from a red-yellow spectrum.
     Hot,
+    /// A palette in which colors are chosen randomly from a green-blue spectrum.
+    ///
+    /// Defaults to
```
Comment (JordiChauzi) [inline src/flamegraph/color/mod.rs]: Is this line correct ?

Comment (jonhoo) [reply src/flamegraph/color/mod.rs]: Nope! Good catch!

File: src/flamegraph/color/mod.rs [hunk 10]
```diff
@@ -16,49 +16,98 @@ const BLUE_GRADIENT: (&str, &str) = ("#eeeeee", "#e0e0ff");
...
     Multi(MultiPalette),
 }
 
+impl Default for Palette {
+    fn default() -> Self {
+        Palette::Basic(BasicPalette::Hot)
+    }
+}
+
+/// A plain color palette in which the color is independent of the function name.
```
Comment (JordiChauzi) [inline src/flamegraph/color/mod.rs]: Actually, the function name may or may not be used to compute the exact color used (depending on the `--hash` option)

**Comment by jasonrhansen** [comment]
Thanks for adding docs for differentials! I made "Flame Graph" the default for the title in `Options`. We should probably document that now.

File: src/flamegraph/color/mod.rs [hunk 11]
```diff
@@ -16,49 +16,98 @@ const BLUE_GRADIENT: (&str, &str) = ("#eeeeee", "#e0e0ff");
...
 #[derive(Clone, Copy, Debug, PartialEq)]
 pub enum BasicPalette {
+    /// A palette in which colors are chosen randomly from a red-yellow spectrum.
```
Comment (JordiChauzi) [inline src/flamegraph/color/mod.rs]: `randomly` <- here again the random aspect is not bound to the palette (only the options `--hash` and/or `--cp`)

File: src/flamegraph/color/mod.rs (hunk 10 above)
Comment (jonhoo) [reply src/flamegraph/color/mod.rs]: Good point. I've tried to address this better in 0676aef07fcb81e210bb143f9586619f14f734c3. What do you think?

**Comment by jonhoo** [comment]
@jasonrhansen done in 924421bff10b4bb0e743bea9543e632e60547ea4

File: src/flamegraph/color/mod.rs (hunk 10 above)
Comment (JordiChauzi) [reply src/flamegraph/color/mod.rs]: Seems good!

**Pull request merged**